Tenant

yasi1989 opened the pull request:
# テナント管理機能の実装
## 変更内容
- テナント管理機能の実装
## 主な変更点
- テナント管理用のデータベースマイグレーションを追加
- 企業情報のCRUD操作APIを実装
- データベーススキーマの更新
- 企業一覧・詳細・登録・編集・削除機能を実装

yasi1989 committed:
fix:テナント取得

yasi1989 committed:
fix:会社コード修正

yasi1989 committed:
fix:merge

yasi1989 committed:
fix:データ

yasi1989 committed:
add:会社管理のロジック追加

yasi1989 committed:
fix:会社削除処理

yasi1989 committed:
fix:不要ファイル削除

yasi1989 committed:
fix:たいぽ

yasi1989 committed:
clean:hooksの処理を綺麗に

yasi1989 committed:
fix:パフォーマンス向上

yasi1989 committed:
fix:コンパイルエラー対応

yasi1989 committed:
chore: mainブランチをマージ

yasi1989 commented on src/features/system/companies/api/actions.ts:
```diff
@@ -0,0 +1,71 @@
+import { revalidatePath } from 'next/cache';
+import { URLS } from '@/consts/urls';
+import { AddCompanySchema, EditCompanySchema } from '../lib/formSchema';
+
+export const addCompanyAction = async (values: z.infer<typeof AddCompanySchema>): Promise<ActionStateResult> => {
+  try {
+    const { companyName, domain } = values;
+    const company = await db.query.companies.findFirst({ where: eq(companies.domain, domain) });
+    if (company) {
+      return { success: false, error: 'ドメインが重複しています。' };
+    }
+    await db.insert(companies).values({
+      companyName,
+      domain,
+    });
+    revalidatePath(URLS.SYSTEM_COMPANIES);
+    return { success: true };
+  } catch (error) {
+    return actionErrorHandler(error);
+  }
+};
```
- 企業の追加機能を実装
- ドメイン重複チェック機能を追加

yasi1989 commented on src/features/system/companies/api/actions.ts:
```diff
@@ -0,0 +1,71 @@
+    return actionErrorHandler(error);
+  }
+};
+
+export const editCompanyAction = async (values: z.infer<typeof EditCompanySchema>): Promise<ActionStateResult> => {
+  try {
+    const { id, companyName, domain } = values;
+    const [currentCompany, existingCompany] = await Promise.all([
+      db.query.companies.findFirst({ where: (companies, { eq }) => eq(companies.id, id) }),
+      db.query.companies.findFirst({
+        where: (companies, { and, eq, ne }) => and(eq(companies.domain, domain), ne(companies.id, id)),
+      }),
+    ]);
+    if (!currentCompany) {
+      return { success: false, error: '会社が見つかりませんでした。' };
+    }
+    if (currentCompany.domain !== domain && existingCompany) {
+      return { success: false, error: 'ドメインが重複しています。' };
+    }
+    await db
+      .update(companies)
+      .set({
+        companyName,
+        domain,
+      })
+      .where(eq(companies.id, id));
+    revalidatePath(URLS.SYSTEM_COMPANIES);
+    return { success: true };
+  } catch (error) {
+    return actionErrorHandler(error);
+  }
+};
```
- 企業の編集機能を実装
- ドメイン重複チェック機能を追加

yasi1989 commented on src/features/system/companies/api/actions.ts:
```diff
@@ -0,0 +1,71 @@
+    return actionErrorHandler(error);
+  }
+};
+
+export const deleteCompanyAction = async (id: string): Promise<ActionStateResult> => {
+  try {
+    const company = await db.query.companies.findFirst({ where: (companies, { eq }) => eq(companies.id, id) });
+    if (!company) {
+      return { success: false, error: '会社が見つかりませんでした。' };
+    }
+    await db.delete(companies).where(eq(companies.id, id));
+    revalidatePath(URLS.SYSTEM_COMPANIES);
+    return { success: true };
+  } catch (error) {
+    return actionErrorHandler(error);
+  }
+};
```
- 企業の削除機能を実装

yasi1989 commented on src/features/system/companies/api/fetches.ts:
```diff
@@ -0,0 +1,15 @@
+import { db } from '@/lib/db/drizzle';
+import { Company } from '@/lib/actionTypes';
+
+export const fetchCompanies = async (): Promise<Company[]> => {
+  try {
+    return await db.query.companies.findMany({
+      orderBy: (companies, { desc }) => [desc(companies.createdAt)],
+    });
+  } catch (error) {
+    console.error('データ取得に失敗しました。', error);
+    throw error;
+  }
+};
```
- 企業の検索機能を実装

yasi1989 commented on src/features/system/companies/hooks/useCompany.ts:
```diff
@@ -1,35 +1,77 @@
-          },
-    resolver: zodResolver(CompanySchema),
+  const form = useForm<z.infer<typeof AddCompanySchema | typeof EditCompanySchema>>({
+    defaultValues: {
+      id: '',
+      companyName: '',
+      domain: '',
+    },
+    values: type === FORM_MODE.EDIT.value && data ? { ...data } : undefined,
+    resolver: zodResolver(type === FORM_MODE.ADD.value ? AddCompanySchema : EditCompanySchema),
   });
-  const onSubmit = (data: z.infer<typeof CompanySchema>) => {
-    startTransition(() => {
-      console.log(data);
+  const onSubmit = (data: z.infer<typeof AddCompanySchema | typeof EditCompanySchema>) => {
+    startTransition(async () => {
+      try {
+        const { success, error } = await actions[type](data);
+        if (!success) {
+          toast.error(`${ERROR_MESSAGE.APPLICATION_ERROR}: ${error}`);
+        } else {
+          toast.success(getFormModeName(type) + 'に成功しました。');
+        }
+      } catch (error) {
+        if (error instanceof Error) {
+          toast.error(`${ERROR_MESSAGE.UNEXPECTED_ERROR}: ${error}`);
+        } else {
+          toast.error(ERROR_MESSAGE.SYSTEM_ERROR);
+        }
+      }
     });
   };
   return { form, onSubmit, isSubmitted };
 };
```
- 企業の追加・編集のカスタムフックを実装
- フォーム管理にreact-hook-formを統合
- バリデーションスキーマを適用
- トースト通知によるユーザーへのフィードバックを実装
- ローディング状態の管理を追加

yasi1989 commented on src/features/system/companies/hooks/useCompany.ts:
```diff
@@ -1,35 +1,77 @@
   return { form, onSubmit, isSubmitted };
 };
+
+export const useDeleteCompany = (id: string) => {
+  const [isSubmitted, startTransition] = useTransition();
+  const onDelete = () => {
+    startTransition(async () => {
+      try {
+        const { success, error } = await deleteCompanyAction(id);
+        if (!success) {
+          toast.error(`${ERROR_MESSAGE.APPLICATION_ERROR}: ${error}`);
+        } else {
+          toast.success('削除に成功しました。');
+        }
+      } catch (error) {
+        if (error instanceof Error) {
+          toast.error(`${ERROR_MESSAGE.UNEXPECTED_ERROR}: ${error}`);
+        } else {
+          toast.error(ERROR_MESSAGE.SYSTEM_ERROR);
+        }
+      }
+    });
+  };
+  return { onDelete, isSubmitted };
+};
```
- 削除のフック

yasi1989 committed:
fix:コンフリクト

yasi1989 committed:
fix:コンフリクトの解消
x:q
:q
quit
''

yasi1989 committed:
feat(auth): 認証ミドルウェアをNext.js 16の仕様に更新

yasi1989 committed:
feat:biome設定の更新と依存関係のアップデート

yasi1989 committed:
refactor: 会社情報の型定義を統一し、プロパティ名を明確化

yasi1989 commented on src/proxy.ts:
```diff
@@ -25,9 +28,7 @@ export default auth((req: NextAuthRequest) => {
-    // Always run for API routes
     '/(api|trpc)(.*)',
   ],
-};
```
- next.js16でmiddlewareがproxyに

yasi1989 committed:
fix:コンパイルエラー対応

yasi1989 committed:
feat:コンパイルエラー

engineeeeeeeeeeeeeeeeeeeer commented on src/features/system/companies/api/actions.ts:
```diff
@@ -0,0 +1,71 @@
+  }
+};
+
+export const deleteCompanyAction = async (id: string): Promise<ActionStateResult> => {
```
スキーマ上は onDelete: 'cascade' が設定されているが、ユーザーへの警告があった方がいいかもと思いました！
onDelete: 'cascade' のため、誤って削除すると、関連する全データ（部署、ユーザー、勤怠、経費など）が一括削除されるので🙇

engineeeeeeeeeeeeeeeeeeeer commented on src/features/system/companies/api/actions.ts:
```diff
@@ -0,0 +1,71 @@
+import { actionErrorHandler } from '@/lib/errorHandler';
+import { AddCompanySchema, EditCompanySchema } from '../lib/formSchema';
+
+export const addCompanyAction = async (values: z.infer<typeof AddCompanySchema>): Promise<ActionStateResult> => {
```
そういえば、認可って実装しているんでしたっけ...??

この辺のAction処理はシステム権限の人だけが操作可能にした方がいいかなと思い、確認でした🙇

engineeeeeeeeeeeeeeeeeeeer commented on src/consts/validate.ts:
```diff
@@ -14,3 +14,9 @@ export const VALIDATION_LIMITS = {
   DOMAIN_REGEX: /^[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9]*\.([a-zA-Z]{2,})+$/,
```
この正規表現だと、.co.jpなどが登録できないので、見直した方がいいかなと思いました！

engineeeeeeeeeeeeeeeeeeeer replied:
<img width="433" height="381" alt="image" src="https://github.com/user-attachments/assets/8026b8f2-9c81-48aa-a1af-b80251e14301" />

engineeeeeeeeeeeeeeeeeeeer reviewed (commented):
いくつかコメントしたので、確認お願いします！

yasi1989 committed:
fix: 権限検証とドメインバリデーションの改善
 
- ロール権限の確認ロジックを修正し、正しいロールIDで検証するように変更
- ドメイン検証の正規表現を更新
  - サブドメインやハイフンを含むドメインを正しく検証できるように修正

yasi1989 commented on src/consts/validate.ts:
```diff
@@ -11,12 +11,12 @@ export const VALIDATION_LIMITS = {
   MAX_FILE_SIZE: MAX_MB * 1024 * 1024,
   ACCEPTED_FILE_TYPES: ['image/jpeg', 'image/jpg', 'image/png', 'image/gif', 'image/webp'],
-  DOMAIN_REGEX: /^[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9]*\.([a-zA-Z]{2,})+$/,
+  DOMAIN_REGEX: /^[a-zA-Z0-9]([a-zA-Z0-9-]*[a-zA-Z0-9])?(\.[a-zA-Z0-9]([a-zA-Z0-9-]*[a-zA-Z0-9])?)*\.[a-zA-Z]{2,}$/,
```
サブドメインやハイフンを含むドメインを正しく検証できるように修正

yasi1989 commented on src/features/auth/lib/authUtils.ts:
```diff
@@ -41,3 +41,17 @@ export const credentialsSignIn = async (email: string, password: string): Promis
   }
 };
+
+export const requireAuth = async () => {
+  const _session = await auth();
+  if (!_session?.user?.id) {
+    throw new Error('認証情報が見つかりませんでした。');
+  }
+  return _session as Session & {
+    user: {
+      id: string;
+      email?: string | null;
+      name?: string | null;
+    };
+  };
+};
```
セッションのユーザ情報の取得処理

yasi1989 commented on src/features/auth/lib/authRoleUtils.ts:
```diff
@@ -0,0 +1,24 @@
+import { requireAuth } from './authUtils';
+import { getRoleById } from './roleCache';
+
+export const requireRole = (allowsRoles: RoleCodeType[]) => async () => {
+  const session = await requireAuth();
+  const user = await db.query.users.findFirst({ where: eq(users.id, session.user.id) });
+  if (!user) {
+    throw new Error('ユーザー情報が見つかりません');
+  }
+  if (!user.roleId) {
+    throw new Error('ユーザーのロールが設定されていません。');
+  }
+  const userRole = await getRoleById(user.roleId);
+  if (!allowsRoles.includes(userRole.roleCode as RoleCodeType)) {
+    throw new Error('権限がありません。');
+  }
+  return { session, user, userRole };
+};
+
+export const requireSystemAdmin = async () => requireRole([ROLE.SYSTEM_ADMIN])();
```
ロール権限のチェック処理

yasi1989 commented on src/features/auth/lib/roleCache.ts:
```diff
@@ -27,3 +27,10 @@ export async function getRoleByCode(roleCode: string): Promise<RoleType> {
   return role;
 }
+
+export async function getRoleById(roleId: string): Promise<RoleType> {
+  const cache = await getRoleCache();
+  const role = Array.from(cache.values()).find((r) => r.id === roleId);
+  if (!role) throw new Error(`Role with id '${roleId}' not found`);
+  return role;
+}
```
ロールIDからロール情報を取得する処理がなかったので追加

yasi1989 commented on src/features/system/companies/components/DeleteCompanyDialog.tsx:
```diff
@@ -7,7 +7,22 @@ type DeleteCompanyDialogProps = {
 const DeleteCompanyDialog = ({ id }: DeleteCompanyDialogProps) => {
   const { onDelete, isSubmitted } = useDeleteCompany(id);
-  return <DeleteButton onDelete={onDelete} isLoading={isSubmitted} />;
+  const _description = (
+    <>
+      <p className="font-semibold block">この操作は取り消すことができません。</p>
+      <div className="block rounded-md bg-destructive/10 p-3 text-sm">
+        <p className="font-semibold mb-2 block">以下のデータが完全に削除されます：</p>
+        <ul className="list-disc list-inside space-y-1">
+          <li>すべての部署情報</li>
+          <li>すべてのユーザーアカウント</li>
+          <li>すべての勤怠記録</li>
+          <li>すべての経費申請</li>
+          <li>その他関連する全データ</li>
+        </ul>
+      </div>
+    </>
+  );
```
削除時に関連情報が削除されてしまうことを警告

yasi1989 reviewed (commented):
fix: 権限検証とドメインバリデーションの改善
 
- ロール権限の確認ロジックを修正し、正しいロールIDで検証するように変更
- ドメイン検証の正規表現を更新
  - サブドメインやハイフンを含むドメインを正しく検証できるように修正

yasi1989 committed:
chore: Biome の設定を更新し、コードフォーマットを一括修正

yasi1989 commented on src/app/(auth)/error/page.tsx:
```diff
@@ -1,6 +1,6 @@
+import { Suspense } from 'react';
 import { AuthErrorContent } from '@/features/auth/error/components/AuthErrorContent';
 import { AuthErrorFallback } from '@/features/auth/error/components/AuthErrorFallback';
-import { Suspense } from 'react';
 
 export default function AuthErrorPage() {
   return (
```
インデント、インポート順などを一括修正

yasi1989 reviewed (commented):
chore: Biome の設定を更新し、コードフォーマットを一括修正

インデントなどの修正のため確認不要です！

yuki-yuki-engineer reviewed (approved):
LGTM!!!
